### DISCUSSION TITLE
Best way to extract the top N keys from a local cache?

### DISCUSSION THREAD

**Discussion by javafanboy** [oracle/coherence#85] [General]
When starting new VMs participating as non-storage enabled nodes in a "near cache" with already several tens of nodes, we would like to pre-load it with the "best" key values based on the other front caches on already running VMs usage statistics (we need to think about if creating some average or trusting the longest running VM etc) and my question is **what the best method would be to getting hold of the "top N" keys from a local cache?**

My plan is to onced I know the top N keys perform a number of getAll calls with the very best keys in the last one (this way they will have a slightly more recent access time making them a little less likely to get evicted). My thinking is also to use an N that is quite a bit smaller than the near cache capacity once again to avoid the "good" keys/values from getting evicted before they have started picking up use and to keep the load time down to get the new VMs into use fast.

I am not aware of any way to otherwise "prime" the usage data of newly loaded keys to make them less likely to get evicted - if anybody have some tips on that let me know!

**Comment by aseovic** [comment]
@javafanboy We don't keep statistics at the key level (and it would be cost prohibitive to do so), but you could possibly use the Invocation Service to get the front cache key set from each of the existing storage disabled members and then preload the intersection of those key sets, or keys that are present on > X% of members, or something similar.

That said, you may not gain much, if anything -- it may be more expensive to calculate and preload the entries that _you think_ are "the best", than to simply load them as the requests on the new member come in. 

Basically, you are optimizing a simple get from a remote member, which is not an expensive operation unless the entry itself is pretty big (I assume that the latency between storage enabled and storage disabled members is quite low). It may be best just to let the near cache do its thing and prime itself based on the actual requests coming in. Just my two cents.

**Reply by javafanboy** [reply]
Thanks for your input. This question is asked from a public cloud perspective where we auto scale non-storage ensbled nodes and want to get the new nodes operating at full speed as quickly as possible.
What we are caching is objects forming a graph (i e objects contain keys to other related objects in the cache) and if not pre-warming the local cache a large number of requests will, when traversing the graph get hundreds or even thousands of cache misses each on individual or a handful (if batching them until further progress is not possible) of objects at a time as they are discovered. Even with 10 GB network the latency of these many cache misses are significant.as they are discovered sequentially.
I was surprised to read that key level statistics would not be kept for local caches - how are then eviction choices made (assuming size limited cache)? I was under the impression that various methods like LRU, MFU and hybrid would require statistics and at least the latest option of using Caffeine as near cache stores key info if in a very compact form but perhaps there is no Coherence mechanism to retreive it?
Anyhow the idea of looking at intersection of key maps sounds interesting and simple to do with some periodicity to produce good pre-load candidates!

**Reply by ben-manes** [reply]
fwiw, Caffeine provides the hottest entries through its `Policy.Eviction` interface. For example,
```java
Map<K, V> top100 = cache.policy().eviction().orElseThrow().hottest(100);
```

This is not exposed by the Coherence adapter.

**Reply by aseovic** [reply]
Sorry, I wasn't clear -- we don't keep per-entry statistics in a central place (such as MBean server) where you could go to get all the info you need at once. We certainly keep things such as touch/access count and last access time locally, at the entry level, and you can access those values via `getTouchCount` and `getLastTouchMillis` methods on the `LocalCache.Entry`.

We will also call `ExpirationPolicy.entryTouched` whenever an entry is touched (if a custom/external expiration policy is provided), so that may give you a way to pre-calculate a set of "hot" keys on each member. You would still need to use the Invocation service to retrieve those keys from each member, in order to come up with a set of keys to pre-load.

Obviously, the above is very implementation dependent, and will only work if `LocalCache` is used as a near cache front map. If you are using Caffeine, you may be able to use the approach @ben-manes suggested above.

At the moment you would have to use reflection to access `CaffeineCache.f_eviction` field, but we _could_ add public accessors for Caffeine objects we wrap, such as `Cache`, `Eviction` and `VarExpiration`.

@ben-manes What do you think? I'm not opposed to the idea...

**Reply by ben-manes** [reply]
I think providing the underlying Cache (where the policies can be fetched) is reasonable. In JCache there is an `unwrap(class)` method if you wanted it on an interface. An implementation-specific method would be my inclination.

**Reply by aseovic** [reply]
Yeah, my only concern is that would open the door for a direct `Cache` mutation, bypassing the "official" `ConfigurableCacheMap` API, which could potentially lead to inconsistencies with cache stats reporting, would not trigger events, etc.

That said, considering the hoops you'd have to jump through to get to the actual instance, it may be beneficial for the advanced use cases, such as this one. In other words, if you access wrapped `Cache` instance directly, you better know what you are doing.

**Reply by ben-manes** [reply]
That’s true, the callbacks and things could be ignored (or only partially applied). Not all of the policy methods are read only, so there would be footguns but certainly expected to be unsafe if used so maybe okay.

**Reply by aseovic** [reply]
@javafanboy Do you have an opinion on the subject? Would exposing Caffeine `Cache` via public API be beneficial enough to warrant doing it, even though it makes it easier for the users such as yourself to shoot themselves in a foot? ;-)

**Reply by javafanboy** [reply]
> @javafanboy Do you have an opinion on the subject? Would exposing Caffeine `Cache` via public API be beneficial enough to warrant doing it, even though it makes it easier for the users such as yourself to shoot themselves in a foot? ;-)

I am happy for all chances to optionally shoot myself in the foot if that provide new useful functionality - i.e. as long as it does not make the most common usage patterns more complex or error-prone it is fine :-)

Easy things should be easy to do while it is ok that advanced things are a bit trickier and even requires more "doing them in the right way" to work properly...

**Reply by aseovic** [reply]
@javafanboy That's how I feel about it as well. This is a fairly advanced feature, and you will have to jump through several hoops to get the backing map, cast it into the correct type, and access the property it provides. Anyone who can figure out how to do all that is likely knowledgable enough not to use that property the wrong way.

Plus, it's not like we could prevent you from accessing it via reflection anyway -- this will just make it simpler and cleaner.

All right, I'll open an issue and make the change. Should be in the next patch release for both 22.06.x and 22.09.x code lines.

**Reply by javafanboy** [reply]
Sounds good - thanks!

**Comment by javafanboy** [comment]
From my personal point of view my previous improvement proposal "Allow auto scaling without making cluster vulnerable" is however much more important as it makes a big difference for Coherence CEs suitability in a (non-Kubernettes) cloud environment.
